### ISSUE TITLE
Improve unpythonic.fix.fix

### ISSUE THREAD

**Issue by Technologicat** [Technologicat/unpythonic#47]
If we ditch the internal memoizer, and use `unpythonic.fun.memoize` instead, `fix` simplifies to:

```python
def fix(bottom=typing.NoReturn, memo=True):
    if bottom is typing.NoReturn or not callable(bottom):
        bottom = const(bottom)
    def decorator(f):
        f_memo = memoize(f) if memo else f  # TODO: thread safety?
        @wraps(f)
        def f_fix(*args, **kwargs):
            e = _get_threadlocals()
            me = (f_fix, args, tuple(sorted(kwargs.items(), key=itemgetter(0))))
            mrproper = not e.visited  # on outermost call, scrub visited clean at exit
            if not e.visited or me not in e.visited:
                try:
                    e.visited.add(me)
                    return f_memo(*args, **kwargs)
                finally:
                    e.visited.clear() if mrproper else e.visited.remove(me)
            else:  # cycle detected
                return bottom(f_fix.__name__, *args, **kwargs)
        f_fix.entrypoint = f  # just for information
        return f_fix
    return decorator
```

This version retains the property that it re-uses known results also in the middle of a call chain.

As for thread safety, `memoize` itself does nothing special, and the memo is shared between threads. At most, this leads to a few spurious overwrites in the memo with new instances of the same value (because `f` is assumed pure), until all threads agree that the key already exists (at which point the memo actually kicks in). The actual dictionary write is one bytecode instruction, so it is atomic. See [SO](https://stackoverflow.com/questions/1312331/using-a-global-dictionary-with-threads-in-python).

The memo lives in the closure of `f_fix`. Exceptions are memoized, too. We call `bottom` at most once for each discovered infinite loop. Even if `bottom` blows up, the exception just ends up in the memo (there is at least one `f_memo` call active when `bottom` is called, because `e.visited` is empty at the start of the call chain).

We no longer need `unwrap`, since there's no need to inspect the return value. Reading the original code again, I'm no longer sure what `n` and the `while` loop are for, as they only surround the outermost call, but at least for `fix` to work as advertised in `unpythonic` they're not needed.

The important property used by `fix` is that **our** call to `f` (or to `f_memo`) will call the original function, **but** *when the original function calls itself, it actually calls our `f_fix`* because name lookup is dynamic.

**TODO**: Merge this in, think a bit more, and then update the docs. Think about TCO.

**Comment by Technologicat** [comment]
**TCO**: In the new version, there's just one place left where we call `f_memo`, so we can insert a trampoline there, and then declare `@fix` as a decorator that applies TCO.

(That's easier for the macro layer, since there's already a categorization between decorators that apply TCO and those that don't, but no category for not-TCO-but-works-together-with-TCO.)

Finally, think of a suitable priority and `regutil` this, to make macros handling decorated lambdas work properly with `fix`.

**Comment by Technologicat** [comment]
Resolved by 5a44811.
